### PR TITLE
validate posted data

### PR DESCRIPTION
**Pull request by dblatcher**
## What does this change?

The '/documents/:id/steps' POST route will validate the body of the request against a model for prosemirror-transform `Step` and return a 400 response to the client if the body is not in the expected format.

The validation is very basic - it only verifies that the request body is an array in which each item is an object with a 'stepType' field with a value matching the `jsonID` property of one of the `Step` subclasses defined in "prosemirror-transform". It does not validate that the objects have other the properties needed to be a valid step of its type.

This seemed a reasonable degree of validation given:
 - we'd want to avoid writing custom validation code that may be too strict
 - best to avoid expensive logic on what will be a very busy endpoint
 - this part of service only needs to store (and retrieve) the serialised step data, not understand them

**note - I'm not sure I'm right about the assumptions above - if we do want tighter validation on incoming data it can be done.**

## How to test

Start the server locally and and go to the 'hello world' index page.
Post data from the  browser console to test the responses eg:

```javascript
async function postSteps(documentId = "", data = {}) {
  const response = await fetch(`/documents/${documentId}/steps`, {
    method: "POST",
    mode: "cors",
    cache: "no-cache",
    credentials: "same-origin",
    headers: {
      "Content-Type": "application/json",
    },
    redirect: "follow", 
    referrerPolicy: "no-referrer", 
    body: JSON.stringify(data),
  });
  console.log('response received:',response.statusText, response.status)
  return response; 
}

postSteps("1234", [42]); // will be rejected
postSteps("1234", [{stepType: 'replace'}, {stepType: 'docAttr'},]); //  will pass
```

## How can we measure success?

This should make the collaboration service more resilient to malformed data being posted to it from Composer (which could potentially happen due to bugs) and make the Composer client able to to log the errors.

## Have we considered potential risks?

We will need to keep the composer edit history service and this validation function in step during development.

### CODE REVIEW

**Review by Fweddi — COMMENTED** [review]
Great work. I've tested it and it works as expected. I agree with keeping things basic for now, we can always add extra requirements later.

I think some of the validation logic is a little complex to understand. It's unclear what a `jsonID` is or does, for example, and `stepTypes` could be renamed to avoid confusion with TS types.

More broadly, could we move towards a more valid type output from these methods, by [parsing, rather than validating](https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/)? If we can pass through some proper typings, we may not need to add the 'valid' property and check for types instead.

**Comment by dblatcher** [comment]
> I think some of the validation logic is a little complex to understand. It's unclear what a `jsonID` is or does, for example, and `stepTypes` could be renamed to avoid confusion with TS types.

Yes, sorry, that is a bit obscure and should have been explained with some comments - will add some.
In brief `jsonID` is property "prosemirror-transform" adds to each subclass of `Step` (eg `ReplaceAroundStep`, `AddMarkStep`). For each subclass, the `jsonID` is the value used to populate the `stepType` when serialising an instance of that subclass with `toJson`.

 
> More broadly, could we move towards a more valid type output from these methods, by [parsing, rather than validating](https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/)? If we can pass through some proper typings, we may not need to add the 'valid' property and check for types instead.

sounds good - I'll look into that.

**Comment by dblatcher** [comment]
I've switched to a parsing pattern and added some typing for Json objects - so our step model is a bit more specific.

File: collab/src/lib/parse-steps.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
+import type { Json } from "../types/json";
+import { stepJsonIds, type StepModel } from "../types/step";
+
+/* eslint-disable -- linter complains of @typescript-eslint/strict-boolean-expressions on line 7. Tried to disable that specific rule but it claims it isn't enabled */
+const isStep = (item: Json): item is StepModel => {
+  return !!(
```
Comment (Fweddi) [inline collab/src/lib/parse-steps.ts:6]: Think we don't need the !! here

File: collab/src/types/step.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+ *
+ * `jsonID` is used internally by `Step.fromJson` to look up the right subclass
+ * to deserialise JSON to.
+ *
+ * We can use it here to derive the list of possible `stepTypes` that a valid
+ * JSON serialised `Step` can have
+ */
+const defaultStepTypes = defaultStepClasses.map(
+  (stepClass) =>
+    (stepClass.prototype as unknown as Step & { jsonID: string }).jsonID
```
Comment (Fweddi) [inline collab/src/types/step.ts:41]: Could we have a separate type for `Step & { jsonID: string }`?
e.g. type StepWithJson = Step & { jsonID: string }
stepClass.prototype as unknown as StepWithJson.jsonID

File: collab/src/types/step.ts [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+/**
+ * Prosemirror allows developer to define custom subclasses of `Step` (see
+ * the annotation on the `Step` type).
+ *
+ * If we do this, we would need to add the `stepType` here.
+ */
+const ourCustomStepTypes: string[] = [];
+
+export const stepJsonIds = [...defaultStepTypes, ...ourCustomStepTypes];
```
Comment (Fweddi) [inline collab/src/types/step.ts:52]: I feel we should be consistent in calling these JsonIds or StepTypes (or maybe a combination, StepTypeIds?)

File: collab/src/types/step.ts (hunk 2 above)
Comment (Fweddi) [reply collab/src/types/step.ts:41]: It seems strange that prosemirror hasn't exported a type for this already!

File: collab/src/types/json.ts [hunk 4]
```diff
@@ -0,0 +1,8 @@
+export type Json =
```
Comment (Fweddi) [inline collab/src/types/json.ts:1]: This seems like a strange type - is it emulating something from prosemirror?

File: collab/src/index.ts [hunk 5]
```diff
@@ -17,12 +19,24 @@ app.get('/healthcheck', (req: Request, res: Response) => {
...
   console.log(
     {
-      requestBody: req.body as unknown,
+      requestBody,
       documentId: req.params['id'],
+      parsedSteps,
     }
   );
+
+  // linter objects to `if (!parsedSteps)` -  @typescript-eslint/strict-boolean-expressions
```
Comment (Fweddi) [inline collab/src/index.ts:33]: Think we don't need this comment?

File: collab/.eslintignore [hunk 6]
```diff
@@ -1 +1,2 @@
 dist/
+coverage/lcov-report/sorter.js
```
Comment (Fweddi) [inline collab/.eslintignore:2]: Worth adding the entire coverage/ directory?

File: collab/src/lib/parse-steps.ts [hunk 7]
```diff
@@ -0,0 +1,26 @@
+import type { Json } from "../types/json";
+import { stepJsonIds, type StepModel } from "../types/step";
+
+/* eslint-disable -- linter complains of @typescript-eslint/strict-boolean-expressions on line 7. Tried to disable that specific rule but it claims it isn't enabled */
```
Comment (Fweddi) [inline collab/src/lib/parse-steps.ts:4]: Think we can ditch this by changing line 7 to `item !== null`.

File: collab/src/types/json.ts [hunk 8]
```diff
@@ -0,0 +1,14 @@
+/**
+ * Recursive type representing any value that can be returned be deserialising a JSON document.
```
Comment (Fweddi) [inline collab/src/types/json.ts:2]: small typo - returned by

File: collab/src/types/json.ts (hunk 4 above)
Comment (Fweddi) [reply collab/src/types/json.ts:7]: Gotcha 👍  This seems a nice lightweight solution for JSON typing

File: collab/src/lib/parse-steps.ts (hunk 1 above)
Comment (Fweddi) [reply collab/src/lib/parse-steps.ts:5]: We still don't need this `!!` I think? The part in brackets will always evaluate to a boolean. `!!` is to coerce types to boolean.

**Review by Fweddi — APPROVED** [review]
A couple of outstanding comments, otherwise this LGTM - great job!

**Pull request merged**